Cleanup | AsyncHelpers Generic State

benrr101 opened the pull request:
## Description
Hello all, in this latest installment, I'm rewriting majority of the AsyncHelper ContinueTask and CreateContinuationTask methods to have generic state. One thing that has bothered me a lot throughout the codebase is that many (if not all) of the ContinueTask or CreateContinuationTask callbacks have lines that decode the state object(s) passed into them. Something like this:

```C#
AsyncHelper.CreateContinuationTaskWithState(
    myTask,
    this,
    (object state) => {
        SqlCommand sqlCommand = (SqlCommand)state;
        sqlCommand.DoSomething();
    });
```

Not only is this challenging to read, it poses the potential for bugs where the type the state object is cast to could be anything, valid or invalid.

Instead, I've taken the liberty to:
1) Rewrite these methods with up to two generics for the stateful methods. This allows full IDE assistance when writing callbacks, improving readability, as well as helping eliminate runtime bugs from incorrect type casts.
2) Clean up the methods to follow a pattern
   1) Records for encapsulating the state object(s) internally to make all task continuations static lambdas
   2) Remove holes in the implemented methods - each method handles the entire continuation (please note that this *does* significantly increase redundant code, but I argue that this is acceptable because chaining the methods together introduces additional layers of call unwrapping and state wrapping, which decreases performance and makes debugging a nightmare - we should aim to remove these methods anyhow and move to async/await natives wherever possible)
3) Removed the "exception converter" from the methods - this was being used around two times and could be handled better by setting the converted exception in the onFailure handler.
4) Clean up calls to these methods
    1) If the method could be moved to using the stateful version, it has been
    2) If the method used a Tuple<x, y> before, it's been decomposed to use the two-generic overload
    3) There are still a handful that could not be migrated to static lambdas due to their excessive amount of closures being used.
5) Add a large number of **unit tests** that test just about every scenario that could happen in the async handler.
    1) This introduces a dependency on Moq

You can now safely rewrite the above example as (generic isn't necessary since it's obvious via object types):
```C#
AsyncHelper.CreateContinuationTaskWithState(
    taskToContinue: myTask,
    state: this,
    onSuccess: this2 => this2.DoSomething());
```

## Issues
N/A

## Testing
Introduced many unit tests for the new code, existing code can be validated via CI.

benrr101 committed:
Extract AsyncHelper from SqlUtil.cs into the utilities namespace

benrr101 committed:
Add generic ContinueTaskWithState - and it's static the whole way through!

benrr101 committed:
Use generic version where not difficult

benrr101 committed:
Add stateful version of CreateContinuationTask

benrr101 committed:
Introduce stateless version that chains into the stateful version

benrr101 committed:
Introduce two-generic ContinueTaskWithState

benrr101 committed:
Introduce two-generic CreateContinuationTaskWithState

benrr101 committed:
Replacing usages that didn't show up before?

benrr101 committed:
Make non-generic ContinueTaskWithState private, remove exteernal usages of it.

benrr101 committed:
Rewrite CreateContinuationTask

benrr101 committed:
Remove seemingly duplicated CreateContinuationTask overload

benrr101 committed:
Remove the non-generic CreateContinuationTaskWithState overload

benrr101 committed:
Rewrite SetTimeoutException

benrr101 committed:
SetTimeoutExceptionWithState

benrr101 committed:
Cleanup WaitForCompletion

benrr101 committed:
Add reference to Moq in unit test project

benrr101 committed:
Adding tests

benrr101 committed:
Removing exception converter from state-less continue task helper

benrr101 committed:
Adding null task cases for CreateContinuationTask

benrr101 committed:
Migrating SqlHelperTest in functional tests to AsyncHelper in unit tests

paulmedynski commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs:
```diff
@@ -2362,15 +2364,15 @@ private Task CopyColumnsAsync(int col, TaskCompletionSource<object> source = nul
                 {
-                    SqlBulkCopy sqlBulkCopy = (SqlBulkCopy)state;
-                    if (i + 1 < sqlBulkCopy._sortedColumnMappings.Count)
+                    if (i + 1 < this2._sortedColumnMappings.Count)
```
Interesting... we're willing to capture `i`, but not `this`?  Could this be:
```
state1: this,
state2: i,
onSuccess: static (connection, count) =>
{
  if (count + 1 < connection._sortedColumnMappings.Count)
   ...
}
```

Hmm, maybe not because of the use of `source` below?

paulmedynski reviewed (commented):
Looked at everything except `AsyncHelper.cs` for now.  Maybe a quick call to explain this a bit more for me (and the team?) would help.

benrr101 committed:
Address copilot comments, address unobserved exception issue that breaks WaitForCompletion test

benrr101 commented on src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.netcore.cs:
```diff
@@ -1334,41 +1334,45 @@ private SqlDataReader GetParameterEncryptionDataReader(out Task returnTask, Task
+            returnTask = AsyncHelper.CreateContinuationTaskWithState(
+                taskToContinue: fetchInputParameterEncryptionInfoTask,
+                state: this,
+                onSuccess: this2 =>
```
I guess I prefer not to since it tidies things up in many cases to one line lambdas. (eg `state => state.Qqq()`)
And the way I look at it, calling it `this2` makes it clear it is `this` just inside the scope.

benrr101 committed:
Merge branch 'main' into dev/russellben/asynchelper-generic-state

benrr101 commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlCommand.Reader.cs:
```diff
@@ -1600,21 +1598,19 @@ private Task RunExecuteReaderTdsSetupContinuation(
-                    ((SqlConnection)state).GetOpenTdsConnection().DecrementAsyncCount();
-                });
+                onFailure: static (this2, _, exception) =>
+                    this2._activeConnection.GetOpenTdsConnection().DecrementAsyncCount());
```
You are right!

benrr101 committed:
Addressing a couple bits of PR comments

paulmedynski commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Utilities/AsyncHelper.cs:
```diff
@@ -0,0 +1,519 @@
+
+namespace Microsoft.Data.SqlClient.Utilities
+{
+    internal static class AsyncHelper
```
What does this class do?  Some docs here and on the methods would be super!

benrr101 replied:
Alrighty, you asked for it - you get comments :)

paulmedynski commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Utilities/AsyncHelper.cs:
```diff
@@ -0,0 +1,519 @@
+        {
+            if (taskToContinue is null)
+            {
+                onSuccess();
```
Interesting choice - if no task to continue was supplied, we consider it to have succeeded, even though it doesn't exist?

benrr101 replied:
This is a remnant of ye olde async/sync code that returns null tasks when executing synchronously. So it's still desired that the onSuccess executes regardless of whether the preceding action was synchronous or asynchronous.

benrr101 replied:
Sure thing

paulmedynski commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/SqlClient/Utilities/AsyncHelperTest.cs:
```diff
@@ -0,0 +1,1291 @@
+        {
+            // Arrange
+            Task taskToContinue = Task.CompletedTask;
+            TaskCompletionSource<object> taskCompletionSource = new();
```
Can you specialize with `int` here instead of `object`, since your result is an integer (one line 34)?

paulmedynski commented on src/Microsoft.Data.SqlClient/tests/UnitTests/Microsoft/Data/SqlClient/Utilities/AsyncHelperTest.cs:
```diff
@@ -0,0 +1,1291 @@
+        private static async Task RunWithTimeout(Task taskToRun, TimeSpan timeout)
+        {
+            Task winner = await Task.WhenAny(taskToRun, Task.Delay(timeout));
+            if (winner != taskToRun)
```
The `Task` class isn't sealed, so it's possible that `taskToRun` is a derived instance that has overridden the `!=` operator to do something other than reference equality.  It's safer here to use `Object.ReferenceEquals(winner, taskToRun)`.

benrr101 replied:
While technically true, we're never using custom task objects in these tests.

mdaigle commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs:
```diff
@@ -2505,53 +2507,53 @@ private Task CopyRowsAsync(int rowsSoFar, int totalRows, CancellationToken cts,
+                                        taskToContinue: readTask,
+                                        taskCompletionSource: source,
+                                        state: this2,
+                                        onSuccess: this3 => this3.CopyRowsAsync(i + 1, totalRows, cts, source));
```
😞

mdaigle commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Utilities/AsyncHelper.cs:
```diff
@@ -0,0 +1,519 @@
+        {
+            try
+            {
+                task.Wait(timeoutInSeconds > 0 ? 1000 * timeoutInSeconds : Timeout.Infinite);
```
Can use TimeSpan instead.

benrr101 commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Utilities/AsyncHelper.cs:
```diff
@@ -0,0 +1,519 @@
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
```
Alrighty... this complicates things a bit, but it'll probably be ok.

Test file has it enabled in the project file

benrr101 committed:
Address PR comments
* Enable nullable
* Add comment blocks
* Use TimeSpan
* Rename state variables to match better

benrr101 committed:
Merge branch 'main' into dev/russellben/asynchelper-generic-state

benrr101 committed:
Increase timeout and ensure no stack dives on task completion

paulmedynski commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Utilities/AsyncHelper.cs:
```diff
@@ -0,0 +1,519 @@
+                        // @TODO: CER Exception Handling was removed here (see GH#3581)
+                        catch (Exception e)
+                        {
+                            taskCompletionSource.SetException(e);
```
Ok - can you make that change throughout?

benrr101 committed:
*Try*Set everything, add comment as per PR comment

paulmedynski commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Utilities/AsyncHelper.cs:
```diff
@@ -297,7 +297,7 @@ internal static void ContinueTaskWithState<TState1, TState2>(
                         catch (Exception e)
                         {
-                            typedState2.TaskCompletionSource.SetException(e);
+                            typedState2.TaskCompletionSource.TrySetException(e);
```
These new TrySetXyz() calls will change the exceptions we could throw (fewer now).  Does that change the behaviour of any of our callers?

benrr101 committed:
Tweak comments, add logging to unovserved exception observation, increase timeout substantially

benrr101 committed:
Fix syntax error

benrr101 committed:
Only trace the event if tracing is enabled (should've read the comments)

benrr101 committed:
Merge branch 'main' into dev/russellben/asynchelper-generic-state

# Conflicts:
#	src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.netcore.cs
#	src/Microsoft.Data.SqlClient/netfx/src/Microsoft.Data.SqlClient.csproj
#	src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlCommand.netfx.cs

benrr101 committed:
Make sure mockOnCancellation is checked on one test, reorder a couple checks b/c I felt like it

benrr101 committed:
Fix tests that don't correspond to their method name
Fix event source listener test so it is ok with the trace from the unobserved exception

benrr101 committed:
Merge branch 'main' into dev/russellben/asynchelper-generic-state

benrr101 committed:
Merge branch 'main' into dev/russellben/asynchelper-generic-state

benrr101 committed:
Hee hee I am dumb.

benrr101 committed:
Fix XEvent test failure :man_facepalming:

benrr101 committed:
Merge branch 'main' into dev/russellben/asynchelper-generic-state

paulmedynski committed:
Merge branch 'main' into dev/russellben/asynchelper-generic-state

paulmedynski committed:
Post-merge tirying-up.

paulmedynski commented:
@benrr101 - These EventSource failures appear to be related to these changes:
```
    Failed Microsoft.Data.SqlClient.ManualTesting.Tests.EventSourceTest.EventSourceTestAll [37 ms]
##[error]EXEC(0,0): Error Message: 
EXEC : error Message:  [D:\a\_work\1\s\build2.proj]
     Assert.Contains() Failure: Item not found in set
  Set:       [0, 1, 2, 3, 4, ┬╖┬╖┬╖]
  Not found: 21
    Stack Trace:
       at Microsoft.Data.SqlClient.ManualTesting.Tests.EventSourceTest.EventSourceTestAll() in D:\a\_work\1\s\src\Microsoft.Data.SqlClient\tests\ManualTests\TracingTests\EventSourceTest.cs:line 36
     at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
     at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

apoorvdeshmukh reviewed (commented):
@benrr101 Can you check the mentioned test failures?
